### ISSUE TITLE
Output index when using HTML5 backend (like a2x)

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#450]
Emulate the behavior of a2x to output an index of index terms in the same way as a2x -f xhtml (via the docbook toolchain). Output in a section with the style `index`, if that section is present.

Example terms:

```
Lions and tigers (Bengal and Siberian) are big cats.
(((Big cats,Lions)))
(((Big cats,Tigers,Bengal Tiger)))
(((Big cats,Tigers,Siberian Tiger)))
Multi-entry terms generate separate index entries.
```

Example index section:

```
[index]
== Index terms
```

Organize the index entries as definition lists and nested index entries as nested definition lists.

**Comment by jxxcarlson** [comment]
I've put code and an example at  [GitHub](https://github.com/jxxcarlson/asciidoctor-make-index) for an experimental utility
program that creates a version of an Asciidoc file with an index 
attached
```
  index_maker: foo.adoc => foo-indexed.adoc
  asciidoctor-latex -b html foo-indexed.adoc =>  foo-indexed.html
```                                                                                                                 
See [this example](https://vschool.s3.amazonaws.com/manuscripts/462.html).  The  index maker code is 
running on [Noteshre.io](http://www.noteshare.io/)

**Comment by vogella** [comment]
AFAICS the index is also not generated if I create PDF output. Is there an similar issue for the PDF output?

**Comment by mojavelinux** [comment]
That's correct, we don't yet output the index in Asciidoctor PDF either. All the information is there, it's just a matter of doing it. I suppose we should add a parallel issue in Asciidoctor PDF to ensure we don't forget.

Keep in mind that it is possible to add the index yourself in the HTML converter using a custom template. You'd need to create a custom file named `section.html.slim` (or `section.html.haml`) and look for the `style == 'index'`. Then, you'd use the `document.references[:indexterms]` table to print out the index.

For a starting point, see https://github.com/asciidoctor/asciidoctor-backends/blob/master/haml/html5/section.html.haml or https://github.com/asciidoctor/asciidoctor-backends/blob/master/slim/html5/section.html.slim.

To do it for Asciidoctor PDF, you'd need to extend the converter and override the `convert_section` method.

It does seem like we should delegate to a dedicated handler for rendering the index content so that overriding the section handler isn't necessary. I think that's a good step to implement as part of this issue.

**Comment by vogella** [comment]
Thanks @mojavelinux for your response, I opened #1661 for the index creation for PDF.

**Comment by mojavelinux** [comment]
I wonder if we should also introduce index as a block macro with the (section) title set using the block title.

```
.Index terms
index::[]
```

That feels more natural and also gives us a place to pass configuration.

**Comment by vogella** [comment]
+1 for the index macro, very consistent

**Comment by vogella** [comment]
+1 for the index macro, very consistent

**Comment by mojavelinux** [comment]
This has now been implemented in the Asciidoctor PDF converter, so it's probably time we do it for the HTML converter as well.

**Comment by mojavelinux** [comment]
```
index::[]
```

The problem with the block macro is that it gets caught inside the previous section, despite the fact that it is intended to create a top-level section. I suppose that's something we could address in the syntax...that certain block macros should terminate sections.

Another idea is to put the macro in the section title:

```
== index::[]
```

That's a little funky, but on the other hand it does communicate the intent quite well.

For the purpose of this issue, I still think we should stick with implementing the original syntax (a special section with style "index"), but I just wanted to note down those ideas.

**Comment by starlinq** [comment]
Is there any progress in index implementation for html5 backend?

**Comment by mojavelinux** [comment]
It's scheduled for the next release. The logic will mirror what was done
for Asciidoctor PDF, so that part is already behind us.

**Comment by bimlas** [comment]
Maybe it should be in another thread, but I have an idea about indexterms: they can be used as "classical" hashtags, but it can be more powerfull because of grouping.

I think a tag icon for each indexterm could be placed on the related block, so the users click on it to jump to the list of related contents. Besides this, it could be more "searchable" if the index list would contain all permutations of indexterms (an option like `:index-permutations:` could enable this).

Here's a real example: http://www.sphinx-doc.org/en/master/genindex.html Search for `coverage_ignore_classes`: it is listed as a standalone entry, having `configuration_value` as a subentry (because the linked "paragraph" related to both terms), besides this it's a subentry of `configuration_value`. Thus the user can found the link easily regardless of the order of the terms.

I think it's a great way to organize knowledge: if I searching for everything that I know about "Quentin Tarantino", then it lists the subelements "Films", "Samuel L. Jackson" and "Music", but if I want to get the list of films of Samuel L. Jackson, then I looking for it's name, "Quentin Tarantino" will be listed under it.

I hope you got the idea: indexterms are great to organize knowledgebase as a blog does it with tags, but it's better because of grouping.

**Comment by jcayouette** [comment]
Excited to see this happen :+1:

**Comment by mojavelinux** [comment]
I've put together a proof of concept you can try today.

https://github.com/asciidoctor/asciidoctor-extensions-lab/pull/92

This prototype helps us understand what will be needed in core to support this natively. The HTML is rudimentary at this point, but it gets us thinking about what we want it to be.

**Comment by mojavelinux** [comment]
@bimlas as far as I understand, indexterms already support grouping in AsciiDoc. As to whether these groups should be created, that's a discussion in #1487. Let's stay focused here on just generating the index from the information we already have.

**Comment by mojavelinux** [comment]
To help move this forward, what I really need is for anyone interested to try out the prototype and let me know whether that is sufficient or, if not, what else is needed. Then we can polish it up and integrate it into core.

**Comment by andrewcarver** [comment]
I'm interested, but not sure I'll catch what's intended, without a sample.adoc.

**Comment by mojavelinux** [comment]
Yep, that works. But what you might want to do to see the references working is to add a bunch of filler between occurrences of the entries.

Triples brackets make a discrete (hidden) index entry. Double brackets make a visible index entry (attaches an anchor to the text).

**Comment by andrewcarver** [comment]
Great, thanks.

**Comment by bimlas** [comment]
It took some times to get it work (never tried Asciidoctor extensions before) so here's the steps for others who interested in trying it out:

* Download https://raw.githubusercontent.com/mojavelinux/asciidoctor-extensions-lab/index-converter-ext/lib/index-converter-ext/extension.rb
* Create a sample document
```asciidoc
= Big cats

Lions and tigers (Bengal and Siberian) are big cats.
(((Big cats,Lions)))
(((Big cats,Tigers,Bengal Tiger)))
(((Big cats,Tigers,Siberian Tiger)))

[index]
== Index terms
```
* Be sure that Asciidoctor is recent enough (`gem install asciidoctor` to upgrade) 
* Run `asciidoctor -r ./extension.rb ./document.adoc`
* You should see something like this:
![image](https://user-images.githubusercontent.com/7034200/39663251-35894312-5070-11e8-8965-7db0a2457039.png)

It seems that the extension is working, but it does not handling multiple occurences of the same term (the very first occurence is linked):

```asciidoc
= Big cats

Lions and tigers (Bengal and Siberian) are big cats.
(((Big cats,Lions)))
(((Big cats,Tigers,Bengal Tiger)))
(((Big cats,Tigers,Siberian Tiger)))

Bengal tigers are living in India.
(((Big cats,Tigers,Bengal Tiger)))

[index]
== Index terms
```

![image](https://user-images.githubusercontent.com/7034200/39663304-2338c54c-5071-11e8-8cdd-d02242a81a75.png)
![image](https://user-images.githubusercontent.com/7034200/39663329-8567f9d6-5071-11e8-93d1-4085f13ad620.png)

Anyway, it's quite usable in my opinion.

**Comment by bimlas** [comment]
Just realized that the max depth of indexterms is limited to 3. Is this intentional?

**Comment by andrewcarver** [comment]
I'm sure that's purely for historical reasons: The original [AsciiDoc indexing-markup](http://asciidoc.org/asciidoc.css-embedded.html#_indexes) was intended only for supporting creation of DocBook `<indexterm>` elements; and in DocBook, the 'max depth of indexterms' (as you put it) was 3.

**Comment by bimlas** [comment]
Anyway, 3 level is enough for most cases in my opinion.

Just tried out the "same indexterm in multiple places" on Asciidoctor-PDF where it works as expected:

![image](https://user-images.githubusercontent.com/7034200/39666778-982a7912-50a9-11e8-9607-1586ef25a0be.png)

**Comment by mojavelinux** [comment]
> but it does not handling multiple occurences of the same term (the very first occurence is linked):

That's correct. The reason is...I'm not sure how we would refer to it. With Asciidoctor PDF, we can use the page number. But we don't really have a way to identify the target in a unique way in HTML. So I'm not really sure what the user would expect to see. Ideas welcome.

**Comment by mojavelinux** [comment]
> Actually, it's NOT PURELY for "historical" reasons, though, is it? Asciidoctor still uses the AsciiDoc indexing-markup to generate the DocBook `<indexterm>` elements

Correct.

And this issue isn't about changing how indexterms work. It's about adding parity with the DocBook toolchain output. If we want to change how indexterms work, that should be the subject of another issue. (And there is already one in play, #1487).

**Comment by mojavelinux** [comment]
> But we don't really have a way to identify the target in a unique way in HTML.

The DocBook toolchain uses the title of the section in which it is contained. But this doesn't really make sense to me as it makes the index hard to read.

One idea would be to append a counter next to the entry. So something like:

```
Big Cats
  Tigers
    Bengal Tiger (1) (2)
```

That seems pretty easy to understand. wdyt?

(we have the information about each target in the model...we just need to decide how to display that information).

**Comment by andrewcarver** [comment]
Before commenting on your helpful suggestion, let's make sure we're quite clear on *why* DocBook XSL people had *such a hard time* coming up with an approach that seemed at all intuitive:

- An index in a hard-copy book is a presentation of **many-to-many** relationships, between topics and the locations (page-nrs) where the topic is discussed; whereas...
- An index made up of hyperlinks (which is all you've got to work with in HTML, since you have no page-nrs) is a presentation of **many-to-one** relationships, between topics and HTML anchor-points that give _some_ indication of location of places where the topic is discussed.

This is why you can't just use the topic-title itself as the hypertext: links in HTML are many-to-one, not multivalued. Ouch.

But the thing about location is also of significance: A researcher wants to be able to CITE some discussion of a particular topic--and the way you must cite is by location! On reflection, this seems to me to be a likely explanation of why the DocBook system settled on section-titles.

However, wouldn't section-NUMBERS serve that purpose, and more cleanly? So, if you do wanna give up the (admittedly unintuitive) approach that the DocBook-to-HTML XSL settled on, there's a suggestion: use section-nrs. (Which is a bit weird unless you include the chapter-nr within the section-nr; which, however, is how Asciidoctor operates, currently.)

Otherwise--i.e., if you don't want to try to use expression of discussion-location--I think what you wanna do here is try to find some kinda hypertext that washes out all semantics, or even the suggestion of semantics, from the actual links. Which, it turns out, is my only, and slight, objection to `[1][2]`--it seems to have just a bit too much whiff of meaning, more than what you intend really to say.

But on the other hand, you're depriving the reader of something useful, if you just go with, say, `[*][*]`: by numbering the links, you're at least giving each one a unique identifier, so the reader could more readily make a note of if they want to go back to it. (Otherwise, they've got to do a count of the `[*]`s appearing prior to the one that led them to a discussion they found interesting and might want to return to.)

So, here's a second suggestion: how about `[a][b][c]`? I like that as it seems to be almost washed clean of any whiff of meaning.

**Comment by andrewcarver** [comment]
P.S. I'm not antagonistic to any of the approaches we've put forward. However, I believe DocBook's folks had a good idea (but didn't express it all that well) by trying, via the section-title, to "give the reader some context for the link." (Stayton, ibid., p. 312) As I said, likely they were trying to give SOME identifier to the discussion-location. But I prefer, for that, to use simply the section-nr., not the whole section-title.

**Comment by oddhack** [comment]
Perhaps something like LaTeX makeidx, where adding an index term can be accompanied by custom text/formatting on the page number (section number, say, in this case).

**Comment by andrewcarver** [comment]
May I make an important clarification, DISTANCING myself from DocBook here in a fundamental way?

In their to-HTML stylesheets, DocBook XSL made a decision which I disagree with **strongly**: taking it as some sort of redundancy error, and having already decided to refer to the section--by title (there being no pg-nrs. to which to refer), they went further and decided to COLLAPSE MULTIPLE OCCURRENCES of the same topic within the same section. Thus, THE LINK NO LONGER GOES TO THE EXACT LOCATION OF THE DISCUSSION being referenced!!?!??!

Here's how Bob Stayton states, describes, and explains this decision they made--which I am repudiating:

- "The links go to the top of the section rather than to the anchor point within the section. That's done to permit multiple identical indexterms in the same section to collapse to a single entry. That's done to avoid having to repeat the section title."

Huh? You decide to take away the user's ability to go right to the location--making them search for it throughout a section which may span SEVERAL PAGES--just so you could avoid some (informative! NOT redundant!) repetition of location-references?? Just so your peculiar sensibilities could not be offended??

No. What I'm suggesting, when I say we take DocBook's approach of using some natural, existing identifier for the discussion's location, is NOT that the links should go "to the top of the section rather than the anchor point within the section." I think it's actually important to RETAIN multiple references to the section, one for each location where that thing was discussed within that section: that lets the reader know HOW MANY discussion-locations that topic has within that section--an important and useful bit of info!

**Comment by andrewcarver** [comment]
P.S. By "no non-silly reason", I did not mean, of course, there are no possible technical difficulties--possibly even insuperable ones--in DocBook XSL's making the links go right to the locations. I ;guess I'll very soon get to discover how difficult it is :-/  I'm encouraged though, that you've seen to have broken that problem's back already, in your above-described work :)

**Comment by andrewcarver** [comment]
Hey, here's a reasonably user-friendly format--combining just good parts of the above:

Big Cats
  Tigers
    Bengal Tiger,  _2.6.1_ (a, b), _5.4_ (a)

**Comment by qasta** [comment]
Hi All. 
Is there any chance to get indexes working for html5? When can it be happened?

**Comment by bimlas** [comment]
@qasta, see the older comments: https://github.com/asciidoctor/asciidoctor/issues/450#issuecomment-386802779

**Comment by qasta** [comment]
@bimlas, thank you.

I tried to use the extension, but got an error: (NoMethodError) undefined method `prepend' for Asciidoctor::Converter::Html5Converter:Class
```
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution generate-html5 of goal org.asciidoctor:asciidoctor-maven-plugin:1.5.6:process-asciidoc failed: (NoMethodError) undefined method `prepend' for Asciidoctor::Converter::Html5Converter:Class
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:145)`
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        ... 20 more
Caused by: org.jruby.exceptions.RaiseException: (NoMethodError) undefined method `prepend' for Asciidoctor::Converter::Html5Converter:Class
        at RUBY.Converter(...doc/src/main/asciidoctor/extension.rb:167)
        at RUBY.Asciidoctor(...doc/src/main/asciidoctor/extension.rb:109)
        at RUBY.(root)(...doc/src/main/asciidoctor/extension.rb:3)
        at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:1040)
        at RUBY.(root)(.../.m2/repository/org/jruby/jruby-complete/1.7.26/jruby-complete-1.7.26.jar!/META-INF/jruby.home/lib/ruby/shared/rubygems/core_ext/kernel_require.rb:1)
        at RUBY.require(.../.m2/repository/org/jruby/jruby-complete/1.7.26/jruby-complete-1.7.26.jar!/META-INF/jruby.home/lib/ruby/shared/rubygems/core_ext/kernel_require.rb:54)
```
I using maven plugin so added this extension to pom.xml (via "require" plugin configuration parameter).
Is this error caused by old version of asciidoctor (plugin I use is v1.5.6) or there is something that can fix this problem now?

**Comment by mlongnecker** [comment]
Hi @mojavelinux, it looks like you were working on this feature in mid-march. Do you anticipate it showing up a new release of asciidoctor soon?

**Comment by mojavelinux** [comment]
It will very likely go into 2.1.0. That release is being held up by the documentation atm.

**Comment by mojavelinux** [comment]
I designed Asciidoctor 2 in such a way that I will be able to introduce this enhancement in a minor release. The next step is to rebase the current work against master, add support for the see/seealso, then add more tests. I don't know when I'll be able to do that work, though it's certainly something I want to make happen as soon as I can. Whether or not this makes 2.1.0 I couldn't tell you.

**Comment by sean-** [comment]
What's the timeline on a release?

**Comment by db48x** [comment]
Is there any reason why this can't be merged yet? You mention that you want to implement support for
the see and seealso attributes, but asciidoctor-pdf doesn't support those either. A few more tests would be nice; if I wrote them would that push this over the line?

**Comment by dunyakirkali** [comment]
Progress update: I opened #4872 to replace the stale #3160 implementation against the current `main` branch.

The PR adds a presentation-neutral index catalog and HTML5 generation for `[index]` sections, including:

- visible and concealed terms
- primary through tertiary nesting
- multiple exact occurrence links
- formatted terms and deterministic sorting
- `see` and `see-also` associations
- default stylesheet support
- stable repeated conversion and unchanged output when no index section exists

Validation is green: 2,869 tests, 59 Cucumber scenarios, maintained RuboCop scope, stylesheet lint, and gem build.
